# FIN 4000 — Chapter 24: Portfolio Performance Evaluation
*Grading a portfolio manager — Sharpe, Treynor, Jensen's alpha, and why they can disagree*

**ANSWER KEY — fully worked, instructor reference**

This notebook is the Python companion to the Chapter 24 Excel template — same data, same formulas, same numbers. Where the Excel workshop has you build a formula in a cell, this notebook has you build it in a line of Python instead.

## Step 1 — Entry Quiz *(3 minutes, individually)*

1. What's the key difference between the Sharpe ratio and the Treynor measure — what kind of risk does each divide by?
2. What does a positive Jensen's alpha tell you about a manager's performance?
3. Why might a manager rank #1 by the Sharpe ratio but not #1 by the Treynor measure?

---

In [1]:
import numpy as np

## Step 2 — Worked Example *(10 minutes, instructor-led)*

A portfolio earned 14% with a standard deviation of 20% and a beta of 1.1. The risk-free rate is 4% and the market returned 11%. Compute the Treynor measure and Jensen's alpha.

In [2]:
r_f, r_m = 0.04, 0.11
rp_A, sigma_A, beta_A = 0.14, 0.20, 1.1

In [3]:
treynor_A = (rp_A - r_f) / beta_A
required_A = r_f + beta_A * (r_m - r_f)
jensen_A = rp_A - required_A
print(f'Treynor={treynor_A:.2%}  Required={required_A:.2%}  Jensen alpha={jensen_A:.2%}')

Treynor=9.09%  Required=11.70%  Jensen alpha=2.30%


*Treynor = (14%−4%)÷1.1 = 9.09%. Required return = 4%+1.1×(11%−4%) = 11.7%. Jensen's alpha = 14%−11.7% = 2.3%. This manager beat their risk-adjusted benchmark by 2.3 percentage points.*

## Step 3 — Pair Work *(12 minutes, with a partner)*

**1.** Compute this portfolio's Sharpe ratio too: (rₚ−rƒ)÷σₚ — the same formula from Chapter 5, applied to a manager's realized return.

In [4]:
sharpe_A = (rp_A - r_f) / sigma_A
print(f'Sharpe (A) = {sharpe_A:.2f}')

Sharpe (A) = 0.50


**2.** A second manager earned 12% with beta 0.8 and standard deviation 25%. Compute their Treynor measure, Jensen's alpha, and Sharpe ratio.

In [5]:
rp_B, sigma_B, beta_B = 0.12, 0.25, 0.8
treynor_B = (rp_B - r_f) / beta_B
required_B = r_f + beta_B * (r_m - r_f)
jensen_B = rp_B - required_B
sharpe_B = (rp_B - r_f) / sigma_B
print(f'Treynor={treynor_B:.2%}  Jensen alpha={jensen_B:.2%}  Sharpe={sharpe_B:.2f}')

Treynor=10.00%  Jensen alpha=2.40%  Sharpe=0.32


**3.** Compare the two managers on all three measures. Do all three measures agree on who performed better?

In [6]:
print(f'{"":10}{"Treynor":>10}{"Jensen":>10}{"Sharpe":>10}')
print(f'{"Manager A":10}{treynor_A:>10.2%}{jensen_A:>10.2%}{sharpe_A:>10.2f}')
print(f'{"Manager B":10}{treynor_B:>10.2%}{jensen_B:>10.2%}{sharpe_B:>10.2f}')
print('\nB wins on Treynor & Jensen; A wins on Sharpe -- a real ranking reversal.')

             Treynor    Jensen    Sharpe
Manager A      9.09%     2.30%      0.50
Manager B     10.00%     2.40%      0.32

B wins on Treynor & Jensen; A wins on Sharpe -- a real ranking reversal.


## Step 4 — Python Pass *(7 minutes)*

- All three performance measures (Sharpe, Treynor, Jensen's alpha), built from return/σ/beta cells for both managers
- A summary block ranking both managers under each measure side by side
- A short flag noting anywhere the rankings disagree

In [7]:
def evaluate_manager(rp, sigma, beta, r_f, r_m):
    treynor = (rp - r_f) / beta
    jensen = rp - (r_f + beta * (r_m - r_f))
    sharpe = (rp - r_f) / sigma
    return {'treynor': treynor, 'jensen': jensen, 'sharpe': sharpe}

A = evaluate_manager(rp_A, sigma_A, beta_A, r_f, r_m)
B = evaluate_manager(rp_B, sigma_B, beta_B, r_f, r_m)
for measure in A:
    winner = 'A' if A[measure] > B[measure] else 'B'
    print(f'{measure:8}: A={A[measure]:.3f}  B={B[measure]:.3f}  -> {winner} wins')

treynor : A=0.091  B=0.100  -> B wins
jensen  : A=0.023  B=0.024  -> B wins
sharpe  : A=0.500  B=0.320  -> A wins


## Step 5 — Discussion *(3-5 minutes, no calculation)*

> A pension fund is evaluating a manager who will run a small slice of a much larger, well-diversified overall portfolio. Should the fund lean on the Treynor measure/Jensen's alpha or the Sharpe ratio to judge that manager? Why?